### DISCUSSION TITLE
Is smooth scrolling not working?

### DISCUSSION THREAD

**Discussion by Hubro** [neovide/neovide#762] [Q&A]
I have installed Neovide 0.7.0 and enabled multigrid. I know multigrid is enabled because swapping windows around in Neovide is now animated. However, smooth scrolling is still not enabled. If I scroll with the mouse, press <C-e>/<C-y> or <C-u>/<C-d>, the window just jumps to the target position as usual.

Should this be working? Do I need any additional configuration?

**Comment by Kethku** [comment]
I have a hunch that the issue is https://github.com/Kethku/neovide/issues/740. I think with winit, the animation is too quick and isn't captured due to the artificially lower framerate. Do you think that could explain what you're seeing?

**Reply by Hubro** [reply]
Animations for the caret and for windows are working just fine though.

How would I go about testing your hunch? :thinking: I just installed Neovide for the first time yesterday, I'm not really familiar with how to debug stuff. Should I try installing an older release?

EDIT: I tried setting `g:neovide_window_scroll_animation_length` from 0.3 to 5, and it still jumps directly to the target position with no animation.

**Reply by Kethku** [reply]
Hmmm. What version of neovim are you using? Are you able to try nightly?

**Reply by Hubro** [reply]
@Kethku neovim neovim 0.4.4, neovide 0.7.0.

Do you mean the nightly of neovim or neovide?

**Reply by Kethku** [reply]
I meant neovim. I would update to a nightly build of neovim and try there. Another thing you could try is bumping the g:neovide_refresh_rate to 150 or something which is likely higher than your monitor, but will at least paper over the bug I was mentioning in #740

**Reply by Kethku** [reply]
Oh also I would seriously consider running on a build from main. Things have improved a TON since the last release

**Reply by Hubro** [reply]
@Kethku I'm now running the latest neovim from Git (0.5.0) and scroll animations now work. They don't work *well* though :sweat_smile: I'd say around half of my viewport movements are animated properly, the other half has a weird stutter where it looks like the animation is starting from the wrong position. Is this a known issue or should I open a bug report?

EDIT: Actually, scrolling works perfectly if I'm using bleeding edge neovim and Neovide 0.7.0. Scrolling only breaks if I'm running bleeding edge Neovide built from main.

**Reply by Kethku** [reply]
Does setting the framerate to some high number effect the scrolling animation?

**Reply by Hubro** [reply]
@Kethku I tried setting the frame rate to 60, 120 and 180 and it didn't make any noticeable difference. I noticed the stuttering in 2 scenarios:

- The caret is placed at the top of the window and you scroll down one line by pressing `<C-e>`
- Any time the window scrolls past a fold

If you are interested, I can make a video showing the glitchy scroll behavior. For now I have reinstalled neovim 0.4.4, as the bleeding edge version would spontaneously lock up indefinitely and consume 100% CPU.

**Reply by Kethku** [reply]
Yes please. That would be helpful

**Reply by Hubro** [reply]
> Yes please. That would be helpful

https://user-images.githubusercontent.com/597206/123865453-3a031100-d92c-11eb-9cab-c02d999e50e4.mp4

As you can see, every time I scroll past a fold, the viewport jumps weirdly. A different, less extreme jump also happens randomly maybe 20-40% of the time, even when no folds are present. The half page jumps (<C-u>/<C-d>) also do this weird jump when it goes over a fold, but it's much harder to tell.

```
~ 
➜ pacman -Q | grep neovide
neovide-git 0.7.0.r66.g2fc74f8-1

~ 
➜ nvim --version  
NVIM v0.5.0-dev+1450-g5aaa1a1c0
Build type: RelWithDebInfo
LuaJIT 2.0.5
Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions     -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security     -fstack-clash-protection -fcf-protection -Wp,-U_FORTIFY_SOURCE -Wp,-D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/tomas/.cache/paru/clone/neovim-git/src/build/config -I/home/tomas/.cache/paru/clone/neovim-git/src/neovim-git/src -I/usr/include -I/home/tomas/.cache/paru/clone/neovim-git/src/build/src/nvim/auto -I/home/tomas/.cache/paru/clone/neovim-git/src/build/include
Compiled by tomas@zbook

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

**Reply by Hubro** [reply]
I tested the same Neovim + Neovide versions (bleeding edge) on my beastly workstation at home, with the Neovide refresh rate set to 240 on a 240 Hz monitor. The random skipping when scrolling single lines does not happen on this machine. After 5 minutes of messing around, scrolling in different documents, it works extremely smoothly and without a single stutter. An absolute joy! I have longed for this experience in Vim for the best part of a decade :grin:

However, when I scroll across a fold, the screen still jumps nauseatingly.

On my laptop, it doesn't matter if I set the refresh rate to 30, 60, 120 or 240, it will randomly stutter when I scroll, perhaps 30-40% of the time. (The laptop has a 60 Hz screen.)

**Reply by Kethku** [reply]
Ok yeah I think this is due to the update loop I use for winit. There is an open issue and I'm planning on addressing it soon TM.

The fold issue is inherent to the hack I use to do the scrolling at all. Unfortunately unless we can convince folks upstream to add a custom tailored api (this maybe possible, they have talked about it in the past) I don't know how likely it will be to fix this. If I had the time, I would go contribute a PR upstream to introduce what we need, but unfortunately I don't have the time to go page in that codebase. Last time I considered it, I found it overwhelming.

So sadly I don't have a good story for you :( it took a lot to get where we are now hahaha and I'm not sure how much the technique can be pushed further.

**Reply by Hubro** [reply]
I realized it's much easier to see the issue if I increase the animation length to something long like 3 seconds. Here is a new recording with `g:neovide_window_scroll_animation_length` set to 3:

https://user-images.githubusercontent.com/597206/123870709-07a8e200-d933-11eb-9bf1-f433be0d72ee.mp4

I press `10<C-e>` a couple of times at the very start of the video to get down to the folds a bit faster. Other than that, I only press `<C-e>` every second or so to scroll downwards. When it comes to the folds, it looks like it simply jumps to the wrong starting position for the animation. For the random line skip thing, it looks like it jumps all the way to the end state of the animation for 1 frame before it starts the animation. It could be that I just don't see this on my workstation because it's rendering at 240 FPS, and 1 frame is not noticeable.

Here is the log that was produced during the above video: https://gist.githubusercontent.com/Hubro/54dabef23cde1c112954a1aace3cabb2/raw/685ec7955d7149d6b82c23ee8a5ba773ed44e2f6/neovide_rCURRENT.log

**Reply by Hubro** [reply]
> Ok yeah I think this is due to the update loop I use for winit. There is an open issue and I'm planning on addressing it soon TM.
> 
> The fold issue is inherent to the hack I use to do the scrolling at all. Unfortunately unless we can convince folks upstream to add a custom tailored api (this maybe possible, they have talked about it in the past) I don't know how likely it will be to fix this. If I had the time, I would go contribute a PR upstream to introduce what we need, but unfortunately I don't have the time to go page in that codebase. Last time I considered it, I found it overwhelming.
> 
> So sadly I don't have a good story for you :( it took a lot to get where we are now hahaha and I'm not sure how much the technique can be pushed further.

Dang, that's too bad. I think I read a comment from one of the containers in the smooth scrolling issue that they wanted this use case to be possible, so hopefully it's something they will consider adding. Have you created a ticket about it on their Github repo? :grin: 

Would it be possible to detect whether you are about to scroll across a fold, and disable the smooth scrolling animation for that movement? That would probably be a lot less jarring than being thrown around :P

**Reply by Hubro** [reply]
@Kethku As soon as I fullscreened Neovide on my workstation, the stutter issue now happens here too :smile: Probably because it's not getting anywhere near 240 FPS anymore, since it's running at 5120x1440. That's probably something that could be optimized :grin: No need to redraw the entire 5120x1440 for every frame when only one pane is being scrolled.

**Reply by adsick** [reply]
can this smooth thing just be working?
I've run 'neovide --multiGrid' and saw no difference

**Reply by Hubro** [reply]
@adsick You need Neovim 0.5.0+ and a nightly build of Neovide. Do you have both of those?

**Reply by adsick** [reply]
> @adsick You need Neovim 0.5.0+ and a nightly build of Neovide. Do you have both of those?

nope, but how did I know that I need it - I've seen no info about it

**Comment by crispinb** [comment]
I'm just trying out neovide (v  0.10.4), on a Linux/Wayland system. Multigrid is animating windows but I don't see smooth scrolling. 

Is this likely to be the same problem ?
